### PR TITLE
[WIP] External chain configuration, chain-specific data directories, configure chains from command line

### PR DESCRIPTION
**Pull request by whilei**
PR/branch addresses the following issues:
- #65 _Configure chains from command line_
- #100 _Use a different data directory than Ethereum_ 
- #131 _Externalizable config for chains_

__Still TODO in the code:__
- [x] Write new unit and integration tests to cover new code
- [x] Check coverage
- [x] Update CI tests; Travis, AppVeyor, bats
- [x] Tackle some `TODO`s and `FIXME`s left in the comments, ie refactoring/logic-proofing opportunities
- [x] Standardize/prettify logging in some places

__TODO otherwise:__
- [ ] Documentation: 
 - Wiki
 - Readme
- [x] Proofread/consensus default names and other hardcoded user-facing values, ie
-  "EthereumClassic" vs "Ethereum-Classic" datadirs 
- "mainnet" vs "mainnet-classic"
- JSON marshal tags (camelCase, hy-phenated, snake_case?)

----

#### There are still some open questions about parts of these changes, including:
 
- [x] Is it OK?/how-to-handle __moving where we store users' data?__ Assuming we can make an automatic and seamless transition between directories within Geth, there is still the liability that we might break external scripts/libraries that depend on the original location. 
> __[base data dir]__ /home/Ethereum ➡️ /home/EthereumClassic
> __[default chain data]__ /home/EthereumClassic/chaindata,nodes,keystore... ➡️ /home/EthereumClassic/mainnet/chaindata,nodes,keystore...

_Possible solutions:_
1. Move it automatically, but provide a disclaimer/warning along with the new release.
2. Move it optionally, either with a `--migrate` flag or interactive prompt or something.
3. Don't move existing data, but new data created would follow new schemes.

----

- [x] Should the __external config file settings be "hard" or "soft"?__ That is to say, if the config file specifies some but not all of the possible configuration fields (`genesis`, `chainconfig`, `name`, etc.), should geth panic, or quietly-but-with-logging fallback to use the default configs in those cases?
> ~Currently implemented is "soft", without any flags. See [this line of code](https://github.com/ethereumproject/go-ethereum/blob/chain-config/cmd/geth/flag.go#L176) for an example.~ changed

_Possible solutions:_
1. Use "hard", assuming that in most cases JSON configuration files will have been created with `dumpExternalChainConfig` command and thus be complete (with good reason to error if they're not), as well as that they're wanting to rely completely on total customization.
2. Use "soft", assuming that users will most generally want to override some but not all of the default configurations; it _should just work_.
3. It doesn't matter, just provide an additional flag (`--soft` or `--hard`) to allow the user to override the default hard/soft setting.

### CODE REVIEW

**Comment by splix** [comment]
> how-to-handle moving where we store users' data? Assuming we can make an automatic and seamless transition between directories within Geth, there is still the liability that we might break external scripts/libraries that depend on the original location.

Move it automatically, but provide a disclaimer/warning along with the new release. We'll write about it in release notes, and a message in console. 

JFYI, party did similar one or two times, and I don't remember any warning message at all, just discovered that my scripts not working properly. Exactly as you warned

--

> Should the external config file settings be "hard" or "soft"? That is to say, if the config file specifies some but not all of the possible configuration fields (genesis, chainconfig, name, etc.), should geth panic, or quietly-but-with-logging fallback to use the default configs in those cases?

Use "hard". If user missed some important configuration, better to panic, rather that waste his time (and probably money) on running incorrect chain

**Comment by splix** [comment]
PR is not finished yet, please do not merge

**Comment by splix** [comment]
* [ ] store configurations for mainnet and testnet as JSON resources within Geth binary
* [ ] when `--chain` specifies a json file (`--chain mychain.json`) read configuration from specified file or exist with error

**Comment by whilei** [comment]
> when --chain specifies a json file (--chain mychain.json) read configuration from specified file or exist with error

Currently in use is `--chainconfig` because `--chain` is used for specifying the name of the chain and data subdirectory (under `datadir`)
`geth --chain my-custom-net` --> `~/EthereumClassic/my-custom-net/chaindata|nodedir|...`

`geth --chain my-custom-net dumpChainConfig` --> `id: "xx", name: "my-custom-net", genesis: {...}, ...` JSON

**Comment by splix** [comment]
I think it's better to put default configs under following path: `geth/config/mainnet.json` instead of current `geth/data/chain-config_mainnet.json`. Can we do that?

File: core/block_validator.go [hunk 1]
```diff
@@ -283,7 +283,7 @@ func CalcDifficulty(config *ChainConfig, time, parentTime uint64, parentNumber,
 		case "ecip1010":
 			if length, ok := f.GetBigInt("length"); ok {
 				explosionBlock := big.NewInt(0).Add(fork.Block, length)
-				if explosionBlock.Cmp(num) < 0 {
+				if num.Cmp(explosionBlock) < 0 {
```
Comment (whilei) [inline core/block_validator.go]: Real glad you caught this. See 012ba0f for test to ensure.

**Pull request merged**